src/rt/minfo.d

```diff
@@ -60,6 +60,339 @@ struct ModuleGroup
      * Throws:
      *  Exception if it fails.
      */
+    version(all)
+    {
+    void sortCtors()
+    {
+        debug(printModuleDependencies)
+        {
+            import core.stdc.stdio : printf;
+            foreach(_m; _modules)
+            {
+                printf("%s%s%s:", _m.name.ptr, (_m.flags & MIstandalone) ? "+".ptr : "".ptr, (_m.flags & (MIctor | MIdtor)) ? "*".ptr : "".ptr);
+                foreach(_i; _m.importedModules)
+                    printf(" %s", _i.name.ptr);
+                printf("\n");
+            }
+        }
+
+        immutable len = _modules.length;
+
+        // This is ugly to say the least. Would be nice to use a binary
+        // search at least instead of linear...
+        int findModule(in ModuleInfo* mi)
+        {
+            foreach (i, m; _modules)
+                if (m is mi) return cast(int)i;
+            return -1;
+        }
+
+        // allocate the two constructor lists. These will be stored for life of
+        // the program, so use malloc.
+        _ctors = (cast(immutable(ModuleInfo)**).malloc(len * (void*).sizeof))[0 .. len];
+        _tlsctors = (cast(immutable(ModuleInfo)**).malloc(len * (void*).sizeof))[0 .. len];
+
+        // Start out doing the shared ctors. The destruction is done in reverse.
+        auto ctors = _ctors;
+        // current element being inserted into ctors list.
+        size_t ctoridx = 0;
+
+        // allocate some stack arrays that will be used throughout the process.
+        ubyte* p = cast(ubyte *)alloca(len * ubyte.sizeof);
+        auto reachable = p[0 .. len];
+
+        p = cast(ubyte *)alloca(len * ubyte.sizeof);
+        auto flags = p[0 .. len];
+
+
+        // use this to hold the error message.
+        string errmsg = null;
+
+        void print(string[] msgs...)
+        {
+            foreach (m; msgs)
+            {
+                // save to the error message. Note that if we are throwing an
+                // exception, we don't care about being careful with using
+                // stack memory. Just use the GC/runtime.
```

Comment:
Yes this will work. Modules are only initialized/sorted after rt_init, i.e. shared libraries aren't constructed in dso_registry until the runtime is initialized.